### PR TITLE
Bring back eval results

### PR DESCRIPTION
**Pull request by Forgind**
~Still working on final testing.~
Tested building a simple project with and without graph and with and without specifying properties/targets/items, and it works.

Tried building a slnf (that should fail if I specify any properties/targets/items) without any of those but with and without graph, and it succeeded.

First commit just reverts the revert.

Second commit attempts to fix it.

Third commit really fixes it.

Fourth commit tweaks a test.

Fifth commit tightens the access to the graph and ensures sln/slnf graph builds aren't affected.

### CODE REVIEW

**Comment by Forgind** [comment]
@dfederm, any chance you could glance at the graph change I made here to make sure it looks legit to you?

**Comment by dfederm** [comment]
> @dfederm, any chance you could glance at the graph change I made here to make sure it looks legit to you?

Looks more or less like what we brainstormed offline. Since there appears to be an e2e test gap, consider running a test build with `/graph` (probably with and without your new feature too?) just to ensure it doesn't blow up.

**Comment by Forgind** [comment]
I manually built my test project, and I got all the properties/items/etc. I wanted regardless of graph 🙂

**Comment by Forgind** [comment]
(Building it normally with and without graph also worked)

File: src/MSBuild/XMake.cs [hunk 1]
```diff
@@ -1427,17 +1532,29 @@ private static void ResetGatheringSwitchesState()
                             {
                                 if (graphBuildOptions != null)
                                 {
-                                    (result, exception) = ExecuteGraphBuild(buildManager, graphBuildRequest);
+                                    graphResult = ExecuteGraphBuild(buildManager, graphBuildRequest);
+                                    ProjectGraphEntryPoint entryPoint = graphBuildRequest.ProjectGraphEntryPoints.First();
```
Comment (rainersigwald) [inline src/MSBuild/XMake.cs:1536]: There should only ever be exactly one of these in this codepath, right?

```suggestion
                                    ProjectGraphEntryPoint entryPoint = graphBuildRequest.ProjectGraphEntryPoints.Single();
```

File: src/MSBuild/XMake.cs [hunk 2]
```diff
@@ -1427,17 +1532,29 @@ private static void ResetGatheringSwitchesState()
...
+                                    if (!entryPoint.GlobalProperties.ContainsKey(PropertyNames.IsGraphBuild))
+                                    {
+                                        entryPoint.GlobalProperties[PropertyNames.IsGraphBuild] = "true";
+                                    }
+
+                                    result = graphResult.ResultsByNode.First(
```
Comment (rainersigwald) [inline src/MSBuild/XMake.cs:1542]: Likewise if there is more than one match here things are broken, right?
```suggestion
                                    result = graphResult.ResultsByNode.Single(
```

File: src/MSBuild/XMake.cs [hunk 3]
```diff
@@ -1427,17 +1532,29 @@ private static void ResetGatheringSwitchesState()
                             {
                                 if (graphBuildOptions != null)
                                 {
-                                    (result, exception) = ExecuteGraphBuild(buildManager, graphBuildRequest);
+                                    graphResult = ExecuteGraphBuild(buildManager, graphBuildRequest);
+                                    ProjectGraphEntryPoint entryPoint = graphBuildRequest.ProjectGraphEntryPoints.First();
+                                    if (!entryPoint.GlobalProperties.ContainsKey(PropertyNames.IsGraphBuild))
+                                    {
+                                        entryPoint.GlobalProperties[PropertyNames.IsGraphBuild] = "true";
+                                    }
```
Comment (rainersigwald) [inline src/MSBuild/XMake.cs:1540]: I remember you fighting with this but I don't remember how we got to this point. Is it right to set it here? Is it right that the graph doesn't set it?

Comment (Forgind) [reply src/MSBuild/XMake.cs:1540]: @dfederm, feel free to opine, but my understanding is that it's not expected to be set for the entry point—the entry point is just created with the set of global properties known by XMake. The graph itself _does_ set it, and that means that when we get the result back, it has that property set as a global property...but then the set of known global properties per the entry point doesn't match the set that comes back from the result, so I'm adjusting that here.

Comment (rainersigwald) [reply src/MSBuild/XMake.cs]: "The entry point doesn't match anything in the actual graph" is the part that feels wrong to me. Also in retrospect `ProjectGraphEntryPoint.GlobalProperties` should maybe be an `IReadOnlyDictionary` instead but alas.

**Review by rainersigwald — APPROVED** [review]

File: src/MSBuild/XMake.cs [hunk 4]
```diff
@@ -1427,17 +1532,37 @@ private static void ResetGatheringSwitchesState()
...
+                                        if (!entryPoint.GlobalProperties.ContainsKey(PropertyNames.IsGraphBuild))
+                                        {
+                                            entryPoint.GlobalProperties[PropertyNames.IsGraphBuild] = "true";
+                                        }
+
+                                        result = graphResult.ResultsByNode.Single(
+                                            nodeResultKvp =>
+                                            nodeResultKvp.Key.ProjectInstance.FullPath.Equals(entryPoint.ProjectFile) &&
+                                            nodeResultKvp.Key.ProjectInstance.GlobalProperties.Count == entryPoint.GlobalProperties.Count &&
+                                            nodeResultKvp.Key.ProjectInstance.GlobalProperties.All(propertyKvp => entryPoint.GlobalProperties[propertyKvp.Key].Equals(propertyKvp.Value)))
```
Comment (ladipro) [inline src/MSBuild/XMake.cs:1549]: The dictionary indexer throws `KeyNotFoundException` if the key is not found. Is it ok to assume that `nodeResultKvp.Key.ProjectInstance.GlobalProperties` and `entryPoint.GlobalProperties` have exactly the same keys here?

Comment (rainersigwald) [reply src/MSBuild/XMake.cs]: It was not. I switched to `TryGetValue`.

before:

```
MSBUILD : error MSB1025: An internal failure occurred while running MSBuild.
This is an unhandled exception in MSBuild -- PLEASE OPEN A BUG AGAINST THE MSBUILD TEAM.
System.Collections.Generic.KeyNotFoundException: The given key 'TargetFramework' was not present in the dictionary.
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at Microsoft.Build.CommandLine.MSBuildApp.<>c__DisplayClass29_0.<BuildProject>b__1(KeyValuePair`2 propertyKvp) in S:\msbuild\src\MSBuild\XMake.cs:line 1548
   at System.Linq.Enumerable.All[TSource](IEnumerable`1 source, Func`2 predicate)
   at Microsoft.Build.CommandLine.MSBuildApp.<>c__DisplayClass29_0.<BuildProject>b__0(KeyValuePair`2 nodeResultKvp) in S:\msbuild\src\MSBuild\XMake.cs:line 1547
   at System.Linq.Enumerable.TryGetSingle[TSource](IEnumerable`1 source, Func`2 predicate, Boolean& found)
   at System.Linq.Enumerable.Single[TSource](IEnumerable`1 source, Func`2 predicate)
   at Microsoft.Build.CommandLine.MSBuildApp.BuildProject(String projectFile, String[] targets, String toolsVersion, Dictionary`2 globalProperties, Dictionary`2 restoreProperties, ILogger[] loggers, LoggerVerbosity verbosity, DistributedLoggerRecord[] distributedLoggerRecords, Int32 cpuCount, Boolean enableNodeReuse, TextWriter preprocessWriter, TextWriter targetsWriter, Boolean detailedSummary, ISet`1 warningsAsErrors, ISet`1 warningsNotAsErrors, ISet`1 warningsAsMessages, Boolean enableRestore, ProfilerLogger profilerLogger, Boolean enableProfiler, Boolean interactive, ProjectIsolationMode isolateProjects, GraphBuildOptions graphBuildOptions, Boolean lowPriority, Boolean question, String[] inputResultsCaches, String outputResultsCache, Boolean saveProjectResult, BuildResult& result, String[] commandLine) in S:\msbuild\src\MSBuild\XMake.cs:line 1545
```

now:

```sh-session
❯ .\.dotnet\dotnet.exe msbuild -bl -graph S:\play\mt_graph\mt_graph.csproj -getTargetResult:Build
{
  "TargetResults": {
    "Build": {
      "Result": "Success",
      "Items": []
    }
  }
}
```

**Review by YuliiaKovalova — APPROVED** [review]

**Review by MichalPavlik — APPROVED** [review]

**Review by rokonec — COMMENTED** [review]
There seems to be same bug we did for TerminalLogger.
See: https://github.com/dotnet/msbuild/pull/9135

**Pull request merged**